Convert BlockDevice to ORPC Servers

ioeddk opened the pull request:
Wrapped the BlockDevice module as ORPC servers. 
Todo:
- Replace the block device IO submission queues with OQueues. 
- Replace all the IO operations submissions (i.e. enqueue operation) with "produce to OQueue". 
- Keep track of the outstanding IOs.

ioeddk committed:
Added read segments async with closure functions to BlockDevice

ioeddk committed:
updated .gitignore to ignore dev container config folder

ioeddk committed:
wrapped the BlockDevice concret struct as an ORPC server

ioeddk committed:
Added function to submit the biorequest that is binded to the block device.

ioeddk committed:
Added makefile config to only test raid

ioeddk committed:
Debugged the flipped Bio submitter. Use the flipped style in RAID1. Test passed.

ioeddk committed:
Wrapped the block devices as servers

arthurp commented on kernel/comps/virtio/src/device/block/server_traits.rs:
```diff
@@ -0,0 +1,106 @@
+}
+
+#[orpc_trait]
+pub trait PageIOObservable {
```
This trait needs to be shared with the one in the PageCache. You will need to move this somewhere that can be used from both the page cache and this crate.

ioeddk replied:
I had to put it at `kernel/comps/virtio/src/device/block/server_traits.rs` so it can be accessed in `kernel/comps/raid/src/server_traits.rs` and `kernel/comps/raid/src/server_traits.rs`. I also renamed it to "BlockIOObservable" since they are at block granularity. I saw you put your `PageIOObservable` under `kernel/src/fs`, but is file under this path accessible by modules in `kernel/comps`? I remember I had some hard time resolving self referencing in the import tree while trying to import across those two modules.

ioeddk replied:
I also feel it would be better to merge them, since the `PageIOObservable` kind of thing would frequently appear if we want to convert more modules to ORPC.

arthurp commented on kernel/comps/virtio/src/device/block/server_traits.rs:
```diff
@@ -0,0 +1,106 @@
+#[orpc_trait]
+pub trait ORPCBio {
+    /// Process a Bio asynchronously. The reply will be sent to [`ORPCBioRequest::reply_handle`].
+    fn process_bio_async(&self, handle: ORPCBioRequest) -> Result<()>;
```
I would rather have this be an OQueue if possible. However, if that requires creating a bunch of thread, then this is fine.
(Note: I used this pattern of an async method in my pagecache for conversion simplicity, I don't actually think it's very good design.)

ioeddk replied:
This function is removed.

arthurp reviewed (commented):
Overall, I think the design is very reasonable. This may already be your plan, but I would like to see some of the `*_blocks_async_*` and similar methods removed and their uses replaced with the ORPC API. Or alternatively, implement the method using the ORPC API. I want to avoid bypassing the ORPC interface as much as possible.

Once the code comes together more, I can do a full code review.

ioeddk committed:
Added LinnOS policy selector

ioeddk committed:
Use range read for weak observer in the selection policy

ioeddk committed:
Implementation for block device, raid 1, and modularized selectors in ORPC style. Have bugs for the packing a producer in OQueues.

ioeddk committed:
updated git ignore

ioeddk committed:
raid1 with orpc compiled but hangs when mounting

ioeddk committed:
Move the handle_requests_observable() function to a anonymous thread in the init

ioeddk committed:
some debug info prints

arthurp commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -53,16 +73,34 @@ impl BlockDevice {
+            move || {
+                // Attach consumer ONCE outside the loop to avoid race condition
+                // where items could be skipped between consumer drop and re-attach
+                let consumer = server_for_oqueue.bio_submission_oqueue().attach_consumer().unwrap();
```
It's slighly better to attach the consumer before you enter the closure at all. So place this `let` between the `{` and the `move || {`. This means that attachment happens before the init function exits at all.

This probably won't fix your issue, but it could theoretically cause a problem.

ioeddk replied:
Resolved in [68ed08f](https://github.com/ldos-project/asterinas/pull/126/commits/68ed08fa743bef70d5ca0aff29793cf2dbca2b04).

ioeddk committed:
clean up

ioeddk committed:
raid1 orpc version is now running. tested, test passed.

ioeddk committed:
temporary fix to switch the initialization sequence to resolve the hanging issue of startign block devices at boot

ioeddk committed:
updated the .gitignore

ioeddk commented on .gitignore:
```diff
@@ -35,3 +35,7 @@ aster-nix-profile-*.svg
+
+.devcontainer/
+
+notes/
```
I didn't know that we could put them under `.git/info/exclude`. I moved those there.  However, it feels like put .devcontainr in .gitignore might be reasonable. Maybe we can add it somewhere else

ioeddk committed:
Clean Up for Merging

ioeddk commented:
Closed due to a branch management accident. Reopened.

ioeddk commented on kernel/comps/virtio/src/device/mod.rs:
```diff
@@ -47,6 +47,12 @@ pub enum VirtioDeviceError {
+    OQueueAttachmentUnsupported,
+    OQueueAttachmentAllocationFailed,
+    ORPCServerPanicked,
+    ORPCServerMissing,
```
Will update error handling later.

ioeddk committed:
Removed commented code

Mounting ExFat file system will cause hanging for an unknown reason. The code commenting off the use of ExFat is removed for a cleaner merge.

arthurp commented on kernel/comps/block/src/bio.rs:
```diff
@@ -177,6 +203,18 @@ pub enum BioEnqueueError {
+            OQueueAttachError::Unsupported { .. } => BioEnqueueError::OQueueAttachmentUnsupported,
+            OQueueAttachError::AllocationFailed { .. } => BioEnqueueError::OQueueAttachmentAllocationFailed,
+        }
+    }
```
Why not: 

```rust
OQueueAttachError(OQueueAttachError)
```

As the enum variant. Then you can just put the original error into that like:

```rust
fn from(err: OQueueAttachError) -> Self {
    Self::OQueueAttachError(err)
}
```

ioeddk replied:
`BioEnqueueError` needs to have trait bound `Clone, Copy, PartialEq, Eq, Debug` implemented, but it's not there in `OQueueAttachError`, so I cannot make it an enum variant.

arthurp replied:
Where is that bound required? It's weird to have that on an error.

arthurp replied:
See [other comment](https://github.com/ldos-project/asterinas/pull/126#discussion_r2692326249)

arthurp commented on kernel/comps/block/src/bio.rs:
```diff
@@ -282,29 +320,52 @@ impl Default for BioWaiter {
+/// FIXME(yingqi): This clone method is added to satisfies the trait bound for the OQueue, because OQueue needs to 
+/// be able to clone elements in the queue in the case there are mutiple observers. However, 
+/// A BIO REQUEST SHOULD NEVER BE CLONED!!!
+// #[derive(Clone)]
```
This problem will go away soon. I'm working on OQueues which have no type bound. It looks like you have already worked around it though. So just remove the derive.

ioeddk replied:
Resolved in 5b97c24

arthurp commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -34,13 +34,33 @@ use crate::{
     device: Arc<DeviceInner>,
     /// The software staging queue.
-    queue: BioRequestSingleQueue,
+    queue: Arc<BioRequestSingleQueue>,
```
Why did this need to be outlined into an Arc?

ioeddk replied:
Yes, I just need a reference to the `num_outstanding_requests()` function of the `BioRequestSingleQueue` which the `BioRequest` is enqueued at the time when the request is completed. This is not an elegant design, but I can't come up with a better solution for now.

arthurp commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -53,16 +73,34 @@ impl BlockDevice {
+        // Thread 2: Handle requests from the OQueue and enqueue them
+        let server_for_oqueue = block_device_server.clone();
+        spawn_thread(server_for_oqueue.clone(), {
+            let consumer = server_for_oqueue.bio_submission_oqueue().attach_consumer().unwrap();
```
I prefer to write this as:

```rust
spawn_thread(server_for_oqueue.clone(), {
    // inside scope where it is needed, and same name
    let block_device_server = block_device_server.clone();
    let consumer = block_device_server.bio_submission_oqueue().attach_consumer().unwrap();
```

ioeddk replied:
Resolved in [b8dce5c](https://github.com/ldos-project/asterinas/pull/126/commits/b8dce5ca2dcd67bfab81d3960b6226b78d697807)

arthurp commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -53,16 +73,34 @@ impl BlockDevice {
+                // where items could be skipped between consumer drop and re-attach
+                loop {
+                    let request = consumer.consume();
+                    server_for_oqueue.queue.enqueue(request).unwrap();
```
Can you use `?`? We try to avoid panicking code like `unwrap`, especially when it's not obvious that it can never happen.

ioeddk replied:
Resolved in [b8dce5c](https://github.com/ldos-project/asterinas/pull/126/commits/b8dce5ca2dcd67bfab81d3960b6226b78d697807)

arthurp commented on kernel/comps/virtio/src/device/mod.rs:
```diff
@@ -47,6 +47,12 @@ pub enum VirtioDeviceError {
+    OQueueAttachmentUnsupported,
+    OQueueAttachmentAllocationFailed,
+    ORPCServerPanicked,
+    ORPCServerMissing,
```
Can you use `OQueueAttachError(OQueueAttachError)` and the same for ORPCError? If that's not possible, let's at least make the names match.

arthurp commented on ostd/src/task/scheduler/mod.rs:
```diff
@@ -24,6 +24,7 @@ use crate::{
 ///
 /// This function can only be called once and must be called during the initialization of kernel.
 pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+
```
Remove random new line.

ioeddk replied:
Removed in 5b97c24

arthurp reviewed (commented):
Getting there. Overall, it makes sense. But there is still clean up to do.

ioeddk commented on kernel/comps/block/src/bio.rs:
```diff
@@ -314,19 +375,42 @@ impl SubmittedBio {
+    }
+
+    pub fn reply(&self) {
+        self.reply_handle.as_ref().unwrap().produce(BlockDeviceCompletionTrace {
```
`reply_handle` is a producer attached to the `bio_completion_oqueue`. It is called in the `handle_irq()` function in `DeviceInner` to get the latency data and insert the statistics to the completion oqueue.

ioeddk committed:
misc updates to reply the pr review

ioeddk commented on kernel/comps/raid/src/lib.rs:
```diff
@@ -68,29 +81,24 @@ impl Raid1Device {
     pub fn handle_requests(&self) {
+        info!("[raid] dequeuing request");
         let request = self.queue.dequeue();
+        info!("[raid] handling requests");
```
Those were debug traces, removed in 5b97c24

ioeddk commented on kernel/comps/raid/src/lib.rs:
```diff
@@ -121,20 +129,23 @@ impl Raid1Device {
-            let member = self.select_read_member(parent.sid_range());
+            info!("[raid] selecting block device");
+            let member = self.selection_policy.select_block_device().unwrap();
+            info!("[raid] selected block device");
```
Those were debug traces, removed in 5b97c24

ioeddk commented on kernel/src/fs/mod.rs:
```diff
@@ -107,30 +115,32 @@ fn setup_raid1_device() -> Result<Arc<Raid1Device>> {
+            .expect("RAID device type mismatch");
+        raid.handle_requests();
     };
     crate::ThreadOptions::new(task_fn).spawn();
```
This part is meant to behave the same as how the `start_block_devive()` function starts the block device. I re-write it bacause `start_block_device()` downcast ot tp VirtIOBlockDevice, but here I need to cast it to Raid1Device. This thread do only one thing, which is to call `handle_requests()` in an infinite loop. I think it is already in the most concise form. Wrapping it with a new server feel a bit redundant.

arthurp replied:
The reason for the server is because eventually everything will be required to be a server and I want to move in that direction. If `Raid1Device` is already a server, then this should be a thread in that. If doing that is not practical, then you can put a TODO. There will be a continued refactor on this.

ioeddk replied:
I put a todo there in 07dd2845990efe7fca109c5e22d2cd64196fa67e

ioeddk committed:
subsequent updates for PR merge

ioeddk commented on kernel/comps/virtio/src/device/block/device.rs:
```diff
@@ -230,6 +277,8 @@ impl DeviceInner {
             complete_request.bio_request.bios().for_each(|bio| {
                 bio.complete(BioStatus::Complete);
+                /// FIXME(yingqi): How to make sure the reply can only be called once? 
+                bio.reply();
```
Actually, do we need to enforce the call-once property here? Analytically, there won't be situition that more than one thread executing the `handle_irq()` on a submitted request, right?

arthurp replied:
I was responding to the TODO. :-) If we can guarantee it only happens once, I'm OK with it. Though having it checked is generally nice.

ioeddk committed:
made formatting to meet the linting requirement

ioeddk committed:
Resolved Linting Issue

ioeddk committed:
Resolve Linting Problem

arthurp commented on kernel/comps/block/src/bio.rs:
```diff
@@ -314,19 +375,42 @@ impl SubmittedBio {
+        self.reply_handle = Some(reply_handle);
+        self.bio_request_single_queue = Some(bio_request_single_queue);
+        self.submission_time = Some(read_monotonic_time());
+    }
```
That's reasonable and less of a concern for me because the time isn't really semantically important (it can just be missing). That is less true of the other fields. I suspect it's not worth it/practical to change it now, but I would like to understand why we used this so we can try to avoid it in the future as we architect the system.

arthurp commented on kernel/comps/block/src/bio.rs:
```diff
@@ -314,19 +370,50 @@ impl SubmittedBio {
+                latency: read_monotonic_time() - self.submission_time.unwrap(),
+                outstanding_requests: self.num_outstanding_requests(),
+            });
+    }
```
This function doesn't actually "reply". It reports the reports the statistics to observers. This difference *does* matter to someone reading this code because it might make them think this call is the operation that marks the completion of the request, but it's not (at least as far as I can tell).

arthurp reviewed (commented):
A couple more things, but I think we are basically there.

ioeddk commented on kernel/src/fs/mod.rs:
```diff
@@ -54,38 +53,38 @@ fn start_block_device(device_name: &str) -> Result<Arc<dyn BlockDevice>> {
+    info!("[raid] initializing RAID-1 device: {:?}", raid1_device_name);
+    if let Err(err) = setup_raid1_device(raid1_device_name) {
+        error!("[raid] failed to setup RAID-1 device: {:?}", err);
     }
```
Starting an ExFat device causes hanging right at boot, for unknown reasons.

ioeddk committed:
Added license

ioeddk committed:
Updated error handling

ioeddk commented on kernel/comps/block/src/bio.rs:
```diff
@@ -282,29 +316,51 @@ impl Default for BioWaiter {
+
+    /// Since SubmittedBio is created after the bio is submitted to some block device, which
+    /// means it definitely has an outstanding request in the request queue of the block device.
+    bio_request_single_queue: Option<Arc<BioRequestSingleQueue>>,
```
Yeah you're wight. The comment is wrong actually. The reason is that the `Bio.enqueue()` function takes something that implements the `BlockDevice` trait. However, only the `BlockDevice` concrete type contains that `BioRequestSingleQueue`. That means we can only take the reference to the `BioRequestSingleQueue` and add it to the `SubmittedBio` in the enqueue function of the BlockDevice concret3 type. So we need to allow the case that the `bio_request_single_queue` field to be `None` in the case it is enqueued by types other than the concret BlockDevice type but who implemented the `BlockDevice` trait.

ioeddk committed:
Clean up for inaccurate comments.

ioeddk committed:
Misc Updates

ioeddk committed:
misc updates

ioeddk committed:
Added annotations to the exfat issue

ioeddk committed:
Formatted code

ioeddk committed:
resolved linting problem

arthurp reviewed (approved):
We got there! Thanks!
